gc.hpp

```diff
@@ -0,0 +1,316 @@
+};
+
+#endif
+
+void sweep()
+{
+  GCObject **current = &start;
+  while (true)
+  {
+    TRACE("LookingAt: %p, Type: %p, Mark: %x \n", current, (*current)->type, (*current)->mark);
+    if ((*current)->mark != timeWalked)
+    {
+      TRACE("Freeing: %p\n", current);
+      auto next = (*current)->next;
+      free(*current);
+      *current = next;
+    }
+    else
+    {
+      current = &(*current)->next;
+    }
+    if (*current == nullptr)
+    {
+      break;
+    }
+  }
+}
+
+void mark(uint8_t d, GCObject *object)
+{
+  if (object == nullptr)
+  {
+    return;
+  }
+  else if (object->mark == timeWalked)
+  {
+    return;
+  }
+  if (d > 0)
+  {
+    TRACE("pointer[%d]: Value: %p Type: %p\n", d, object, object->type);
+  }
+  object->mark = timeWalked;
+  for (size_t index = 0; object->type[index] != 0; index++)
+  {
+    mark(d + 1,
+         *reinterpret_cast<GCObject **>(reinterpret_cast<uint8_t *>(object) + object->type[index]));
+  }
+}
+
+template <typename ListStyle> struct PointersLinkedListIterator
+{
+  const Pointers<0, ListStyle> *curr;
+  uint32_t index;
+
+  PointersLinkedListIterator()
+  {
+    curr = nullptr;
+  }
+
+  PointersLinkedListIterator(void *start)
+  {
+    index = 0;
+    curr = reinterpret_cast<Pointers<0, ListStyle> *>(start);
+
+    /*
+    TRACE("Size: %u Depth: %u\n", curr->size, curr->depth);
+    for (uint32_t size = 0; size < curr->size; size++) {
+
+      TRACE("pointer[%d]: Value: %p\n", size, curr->values[size]);
+    }
+    */
+  }
+
+  PointersLinkedListIterator &operator++()
+  {
+    index += 1;
+    if (index == curr->size)
+    {
+      TRACE("check: %u %p\n", curr->previous, ListStyle::get(curr->previous));
+      if (ListStyle::invalid(curr->previous))
+      {
+        curr = nullptr;
+      }
+      else
+      {
+        index = 0;
+        curr = reinterpret_cast<const Pointers<0, ListStyle> *>(ListStyle::get(curr->previous));
+        TRACE("Size: %u Depth: %u\n", curr->size, curr->depth);
+      }
+    }
+    return *this;
+  }
+
+  bool operator!=(PointersLinkedListIterator iter) const
+  {
+    return curr != iter.curr;
+  }
+  GCObject *operator*() const
+  {
+    return curr->values[index];
+  }
+
+  PointersLinkedListIterator &begin()
+  {
+    return *this;
+  }
+
+  PointersLinkedListIterator end()
+  {
+    return std::move(PointersLinkedListIterator());
+  }
+};
+
+#if __has_feature(capabilities)
+struct CheriStackIterator
+{
+  void **location;
+
+  bool is_stack(void *ptr)
+  {
+    return cheri_is_address_inbounds(bottom, cheri_getaddress(ptr));
+  }
+
+  bool is_exec(void *ptr)
+  {
+    return ((cheri_getperm(ptr) & 0b10) >> 1);
+  }
+
+  CheriStackIterator()
+  {
+    location = reinterpret_cast<void **>(bottom);
+  }
+
+  CheriStackIterator(void *start)
+  {
+    location = reinterpret_cast<void **>(start);
+  }
+
+  CheriStackIterator &operator++()
+  {
+    location -= 1;
+    while (!cheri_gettag(*location) and (is_stack(*location) or is_exec(*location)))
```

Comment:
This clause looks a bit dangerous to me but I might be wrong. I assume that `is_stack` says "is the address at location part of the stack or not?" ? If so, I expected something like this:

```c
while (is_stack(location) && !cheri_gettag(*location) { ... }
```

Writing that out makes me realise that I'm not sure what the `is_exec` check is doing either.

Reply:
The reason I first check is to see first if the tag is valid and then to check if the pointer is on the stack, as pointers that are on the stack, are of no interest of us, but it is easyer to first just test if the thing is a pointer.

Reply:
Ok, so having `cheri_gettag` first makes sense: but what's `is_exec` doing?

Reply:
It is meant to skip the return addresses that are stored on the stack.

Reply:
Why do we want to skip them? [It doesn't look to me like it skips them anyway! It looks like we scan them?]